updates: add ACIs for RBCD self-management

abbra opened the pull request:
Somehow, I forgot this update file...

Fixes: https://pagure.io/freeipa/issue/9354

abbra commented:
rebased to pick up doc changes

rcritten commented on install/updates/73-service-rbcd.update:
```diff
@@ -0,0 +1,5 @@
+add:aci: (targetattr = "objectclass")(targattrfilters="add=objectclass:(objectclass=resourcedelegation)")(version 3.0;acl "RBCD:Kerberos principals can manage resource-based constrained delegation for themselves";allow (write) userdn = "ldap:///self";)
+add:aci: (targetattr = "objectclass")(targattrfilters="add=objectclass:(objectclass=resourcedelegation)")(version 3.0;acl "RBCD:Managing principals can manage resource-based constrained delegation for other principals";allow (write) userattr = "managedby#GROUPDN" or userattr = "managedby#USERDN";)
+add:aci: (targetattr = "objectclass")(targattrfilters="add=objectclass:(objectclass=resourcedelegation)")(version 3.0;acl "RBCD:Delegated permission to manage resource-based constrained delegation for other principals";allow (write) userattr="ipaAllowedToPerform;write_delegation#GROUPDN" or userattr="ipaAllowedToPerform;write_delegation#USERDN" ;)
+
```
I'm not sure about the RBCD prefix. Typically this is "permission". I guess it does make it clear what it is used for but it doesn't follow convention (though this could be a new convention). Similarly, self-service permissions generally use the selfservice prefix.

I can't recall the specifics of delegation but only writing the objectclass?

abbra replied:
> I can't recall the specifics of delegation but only writing the objectclass?

writing to the objectclass is needed for being able to dynamically add `memberprincipal` attribute for delegation.
The rest is already handled the `ipaallowedoperations` / `ipaallowedtoperform` machinery: you can add a `memberprincipal` value if you have been allowed to operate on this entry and in order to add `memberprincipal` attribute you need to have the `resourcedelegation` objectclass.

Prefixes: may be you are right, I should prefix with `permission:` too.

abbra replied:
We don't always use `permission:` prefix either -- there are plenty recent ACIs that simply have no prefix at all. See, for example, `install/updates/73-subid.update`. It also gives the most complex ACI with target attribute filters...

rcritten replied:
I think that subid is done incorrectly. The update installs the aci into cn=subids,cn=accounts. The subid plugin itself auto-registers permissions. Since those are in cn=permissions,cn=pbac,$SUFFIX they can be found by permission-find, but the ACIs are not there. Presumably a user that wanted to modify it would find it did nothing, but I didn't try that.
I don't have the answer. Your PR will work in that aci's will most likely work, they just aren't visible to users. If that's ok then I'm fine with it.

rcritten commented:
I used the design doc cli examples for testing.

Probably unrelated but trying to remove an non-existent delegation returns a strange error:
`ipa: ERROR: memberprincipal does not contain 'one or more values to remove`

I can't create a delegation as the host (both principals exist and I can add this as admin):
```
[root@ipa rpms]# ipa service-add-delegation cifs/ipa.example.test HTTP/web-service.example.test
ipa: ERROR: Insufficient access: Insufficient 'write' privilege to the 'objectClass' attribute of entry 'krbprincipalname=cifs/ipa.example.test@example.test,cn=services,cn=accounts,dc=example,dc=test'.
[root@ipa rpms]# klist
Ticket cache: KEYRING:persistent:0:krb_ccache_cYzymyY
Default principal: host/ipa.example.test@EXAMPLE.TEST

Valid starting       Expires              Service principal
05/05/2023 15:26:12  05/06/2023 15:19:25  HTTP/ipa.example.test@EXAMPLE.TEST
05/05/2023 15:26:09  05/06/2023 15:19:25  krbtgt/EXAMPLE.TEST@EXAMPLE.TEST
```
Group delegation isn't working:
```
ipa group-add storage-admins
ipa group-add-member --user tuser1 storage-admins
kinit tuser1
ipa service-add-delegation cifs/file.example.test HTTP/web-service.example.test
ipa: ERROR: Insufficient access: Insufficient 'write' privilege to the 'memberPrincipal' attribute of entry 'krbprincipalname=cifs/file.example.test@example.test,cn=services,cn=accounts,dc=example,dc=test'.
```
I didn't test RBCD operationally since this PR is just about ACIs.

I'm not sure if a host permission can manage delegations. I had one failure but I only tried once. I tried to stick with the doc.

I confirmed that the subid acis aren't visible via the permission plugin, like the RBCD. That may be ok. It makes some sense performance-wise to have per container ACIs and these are complex enough I don't think we want users playing with them. So perhaps it is not a problem with the prefix since the permission and aci plugins will never see them.

rcritten commented:
I may well be testing the wrong thing here but I don't really understand what the acis are granting, or what adding an objectclass grants permission for.

jrisc commented:
I tried to apply this update AFTER allowing the `testadmin` user to add delegation rules:
```
# kinit admin
# ipa service-allow-add-delegation test/master.ipa.test --users=testadmin
# kdestroy -A
# kinit testadmin
# ipa service-add-delegation test/master.ipa.test test/cli.ipa.test
ipa: ERROR: Insufficient access: Insufficient 'write' privilege to the 'objectClass' attribute of entry 'krbprincipalname=test/master.ipa.test@ipa.test,cn=services,cn=accounts,dc=ipa,dc=test'.
```
It results in a similar error to the one Rob reported for groups:
```
# wget -P /usr/share/ipa/updates/ https://raw.githubusercontent.com/abbra/freeipa/rbcd-updates/install/updates/73-service-rbcd.update
# ipa-server-upgrade
# ipa service-add-delegation test/master.ipa.test test/cli.ipa.test
ipa: ERROR: Insufficient access: Insufficient 'write' privilege to the 'memberPrincipal' attribute of entry 'krbprincipalname=test/master.ipa.test@ipa.test,cn=services,cn=accounts,dc=ipa,dc=test'.
```

abbra commented:
@rcritten @jrisc I updated the ACIs. All the use cases you tried work for me now, could you please retry?

rcritten commented:
Group delegation as a user is still not working. I get the same error:
```

ipa group-add storage-admins
ipa group-add-member --user tuser1 storage-admins
kinit tuser1
ipa service-add-delegation cifs/file.example.test HTTP/web-service.example.test
ipa: ERROR: Insufficient access: Insufficient 'write' privilege to the 'memberPrincipal' attribute of entry 
'krbprincipalname=cifs/file.example.test@example.test,cn=services,cn=accounts,dc=example,dc=test'.
```

abbra commented:
I tried this opion and it worked for me. Can you show me access/errors log entries for this request?

rcritten commented:
Apache just logs the failure

ldap.INSUFFICIENT_ACCESS: {'msgtype': 103, 'msgid': 7, 'result': 50, 'desc': 'Insufficient access', 'ctrls': [], 'info': "Insufficient 'write' privilege to the 'memberPrincipal' attribute of entry 'krbprincipalname=cifs/vm-10-0-185-32.hosted.upshift.rdu2.redhat.com@example.test,cn=services,cn=accounts,dc=example,dc=test'.\\n"}
ipalib.errors.ACIError: Insufficient access: Insufficient 'write' privilege to the 'memberPrincipal' attribute of entry 'krbprincipalname=cifs/vm-10-0-185-32.hosted.upshift.rdu2.redhat.com@example.test,cn=services,cn=accounts,dc=example,dc=test'.
ipa: INFO: [jsonserver_session] tuser1@EXAMPLE.TEST: service_add_delegation/1('cifs/vm-10-0-185-32.hosted.upshift.rdu2.redhat.com', ['HTTP/web-service.example.test'], version='2.253'): ACIError

I enabled ACI summary logging in 389-ds and it reported

[01/Sep/2023:10:25:17.331700439 -0400] - DEBUG - NSACLPlugin - acl_access_allowed - #### conn=59 op=1 binddn="uid=tuser1,cn=users,cn=accounts,dc=example,dc=test"
[01/Sep/2023:10:25:17.332699521 -0400] - DEBUG - NSACLPlugin - print_access_control_summary - conn=59 op=1 (main): Allow search on entry(cn=ipaconfig,cn=etc,dc=example,dc=test).attr(objectClass) to uid=tuser1,cn=users,cn=accounts,dc=example,dc=test: allowed by aci(21): aciname= "Anonymous read access to containers", acidn="dc=example,dc=test"
[01/Sep/2023:10:25:17.333251286 -0400] - DEBUG - NSACLPlugin - print_access_control_summary - conn=59 op=1 (main): Allow read on entry(cn=ipaconfig,cn=etc,dc=example,dc=test).attr(objectClass) to uid=tuser1,cn=users,cn=accounts,dc=example,dc=test: allowed by aci(21): aciname= "Anonymous read access to containers", acidn="dc=example,dc=test"
[01/Sep/2023:10:25:17.333846238 -0400] - DEBUG - NSACLPlugin - print_access_control_summary - conn=59 op=1 (main): Allow read on entry(cn=ipaconfig,cn=etc,dc=example,dc=test).attr(ipaUserSearchFields) to uid=tuser1,cn=users,cn=accounts,dc=example,dc=test: allowed by aci(230): aciname= "permission:System: Read Global Configuration", acidn="cn=ipaconfig,cn=etc,dc=example,dc=test"
[01/Sep/2023:10:25:17.334249076 -0400] - DEBUG - NSACLPlugin - print_access_control_summary - conn=59 op=1 (main): Allow read on entry(cn=ipaconfig,cn=etc,dc=example,dc=test).attr(ipaGroupSearchFields) to uid=tuser1,cn=users,cn=accounts,dc=example,dc=test: cached allow by aci(230)
[01/Sep/2023:10:25:17.334639703 -0400] - DEBUG - NSACLPlugin - print_access_control_summary - conn=59 op=1 (main): Allow read on entry(cn=ipaconfig,cn=etc,dc=example,dc=test).attr(ipaSearchTimeLimit) to uid=tuser1,cn=users,cn=accounts,dc=example,dc=test: cached allow by aci(230)
[01/Sep/2023:10:25:17.335103039 -0400] - DEBUG - NSACLPlugin - print_access_control_summary - conn=59 op=1 (main): Allow read on entry(cn=ipaconfig,cn=etc,dc=example,dc=test).attr(ipaSearchRecordsLimit) to uid=tuser1,cn=users,cn=accounts,dc=example,dc=test: cached allow by aci(230)
[01/Sep/2023:10:25:17.335499858 -0400] - DEBUG - NSACLPlugin - print_access_control_summary - conn=59 op=1 (main): Allow read on entry(cn=ipaconfig,cn=etc,dc=example,dc=test).attr(ipaHomesRootDir) to uid=tuser1,cn=users,cn=accounts,dc=example,dc=test: cached allow by aci(230)
[01/Sep/2023:10:25:17.335828875 -0400] - DEBUG - NSACLPlugin - print_access_control_summary - conn=59 op=1 (main): Allow read on entry(cn=ipaconfig,cn=etc,dc=example,dc=test).attr(ipaDefaultLoginShell) to uid=tuser1,cn=users,cn=accounts,dc=example,dc=test: cached allow by aci(230)
[01/Sep/2023:10:25:17.336221117 -0400] - DEBUG - NSACLPlugin - print_access_control_summary - conn=59 op=1 (main): Allow read on entry(cn=ipaconfig,cn=etc,dc=example,dc=test).attr(ipaDefaultPrimaryGroup) to uid=tuser1,cn=users,cn=accounts,dc=example,dc=test: cached allow by aci(230)
[01/Sep/2023:10:25:17.336702543 -0400] - DEBUG - NSACLPlugin - print_access_control_summary - conn=59 op=1 (main): Allow read on entry(cn=ipaconfig,cn=etc,dc=example,dc=test).attr(ipaMaxUsernameLength) to uid=tuser1,cn=users,cn=accounts,dc=example,dc=test: cached allow by aci(230)
[01/Sep/2023:10:25:17.337122753 -0400] - DEBUG - NSACLPlugin - print_access_control_summary - conn=59 op=1 (main): Allow read on entry(cn=ipaconfig,cn=etc,dc=example,dc=test).attr(ipaMaxHostnameLength) to uid=tuser1,cn=users,cn=accounts,dc=example,dc=test: cached allow by aci(230)
[01/Sep/2023:10:25:17.337544840 -0400] - DEBUG - NSACLPlugin - print_access_control_summary - conn=59 op=1 (main): Allow read on entry(cn=ipaconfig,cn=etc,dc=example,dc=test).attr(ipaPwdExpAdvNotify) to uid=tuser1,cn=users,cn=accounts,dc=example,dc=test: cached allow by aci(230)
[01/Sep/2023:10:25:17.337892234 -0400] - DEBUG - NSACLPlugin - print_access_control_summary - conn=59 op=1 (main): Allow read on entry(cn=ipaconfig,cn=etc,dc=example,dc=test).attr(ipaGroupObjectClasses) to uid=tuser1,cn=users,cn=accounts,dc=example,dc=test: cached allow by aci(230)
[01/Sep/2023:10:25:17.338236451 -0400] - DEBUG - NSACLPlugin - print_access_control_summary - conn=59 op=1 (main): Allow read on entry(cn=ipaconfig,cn=etc,dc=example,dc=test).attr(ipaUserObjectClasses) to uid=tuser1,cn=users,cn=accounts,dc=example,dc=test: cached allow by aci(230)
[01/Sep/2023:10:25:17.338567508 -0400] - DEBUG - NSACLPlugin - print_access_control_summary - conn=59 op=1 (main): Allow read on entry(cn=ipaconfig,cn=etc,dc=example,dc=test).attr(ipaDefaultEmailDomain) to uid=tuser1,cn=users,cn=accounts,dc=example,dc=test: cached allow by aci(230)
[01/Sep/2023:10:25:17.338910681 -0400] - DEBUG - NSACLPlugin - print_access_control_summary - conn=59 op=1 (main): Allow read on entry(cn=ipaconfig,cn=etc,dc=example,dc=test).attr(ipaMigrationEnabled) to uid=tuser1,cn=users,cn=accounts,dc=example,dc=test: cached allow by aci(230)
[01/Sep/2023:10:25:17.339409528 -0400] - DEBUG - NSACLPlugin - print_access_control_summary - conn=59 op=1 (main): Allow read on entry(cn=ipaconfig,cn=etc,dc=example,dc=test).attr(ipaConfigString) to uid=tuser1,cn=users,cn=accounts,dc=example,dc=test: cached allow by aci(230)
[01/Sep/2023:10:25:17.339756229 -0400] - DEBUG - NSACLPlugin - print_access_control_summary - conn=59 op=1 (main): Allow read on entry(cn=ipaconfig,cn=etc,dc=example,dc=test).attr(ipaSELinuxUserMapOrder) to uid=tuser1,cn=users,cn=accounts,dc=example,dc=test: cached allow by aci(230)
[01/Sep/2023:10:25:17.340130364 -0400] - DEBUG - NSACLPlugin - print_access_control_summary - conn=59 op=1 (main): Allow read on entry(cn=ipaconfig,cn=etc,dc=example,dc=test).attr(ipaSELinuxUserMapDefault) to uid=tuser1,cn=users,cn=accounts,dc=example,dc=test: cached allow by aci(230)
[01/Sep/2023:10:25:17.340496718 -0400] - DEBUG - NSACLPlugin - print_access_control_summary - conn=59 op=1 (main): Allow read on entry(cn=ipaconfig,cn=etc,dc=example,dc=test).attr(cn) to uid=tuser1,cn=users,cn=accounts,dc=example,dc=test: cached allow by aci(21)
[01/Sep/2023:10:25:17.340855726 -0400] - DEBUG - NSACLPlugin - print_access_control_summary - conn=59 op=1 (main): Allow read on entry(cn=ipaconfig,cn=etc,dc=example,dc=test).attr(ipaCertificateSubjectBase) to uid=tuser1,cn=users,cn=accounts,dc=example,dc=test: cached allow by aci(230)
[01/Sep/2023:10:25:17.341219454 -0400] - DEBUG - NSACLPlugin - print_access_control_summary - conn=59 op=1 (main): Allow read on entry(cn=ipaconfig,cn=etc,dc=example,dc=test).attr(ipaKrbAuthzData) to uid=tuser1,cn=users,cn=accounts,dc=example,dc=test: cached allow by aci(230)
[01/Sep/2023:10:25:17.344120367 -0400] - DEBUG - NSACLPlugin - acl_access_allowed - #### conn=59 op=4 binddn="uid=tuser1,cn=users,cn=accounts,dc=example,dc=test"
[01/Sep/2023:10:25:17.344751521 -0400] - DEBUG - NSACLPlugin - print_access_control_summary - conn=59 op=4 (main): Allow search on entry(krbprincipalname=cifs/vm-10-0-185-32.hosted.upshift.rdu2.redhat.com@example.test,cn=services,cn=accounts,dc=example,dc=test).attr(objectClass) to uid=tuser1,cn=users,cn=accounts,dc=example,dc=test: allowed by aci(121): aciname= "permission:System: Read POSIX details of SMB services", acidn="cn=services,cn=accounts,dc=example,dc=test"
[01/Sep/2023:10:25:17.345497513 -0400] - DEBUG - NSACLPlugin - print_access_control_summary - conn=59 op=4 (main): Allow read on entry(krbprincipalname=cifs/vm-10-0-185-32.hosted.upshift.rdu2.redhat.com@example.test,cn=services,cn=accounts,dc=example,dc=test).attr(krbCanonicalName) to uid=tuser1,cn=users,cn=accounts,dc=example,dc=test: allowed by aci(122): aciname= "permission:System: Read Services", acidn="cn=services,cn=accounts,dc=example,dc=test"
[01/Sep/2023:10:25:17.345902026 -0400] - DEBUG - NSACLPlugin - print_access_control_summary - conn=59 op=4 (main): Allow read on entry(krbprincipalname=cifs/vm-10-0-185-32.hosted.upshift.rdu2.redhat.com@example.test,cn=services,cn=accounts,dc=example,dc=test).attr(objectClass) to uid=tuser1,cn=users,cn=accounts,dc=example,dc=test: allowed by aci(121): aciname= "permission:System: Read POSIX details of SMB services", acidn="cn=services,cn=accounts,dc=example,dc=test"
[01/Sep/2023:10:25:17.347280235 -0400] - DEBUG - NSACLPlugin - acl_access_allowed - #### conn=59 op=5 binddn="uid=tuser1,cn=users,cn=accounts,dc=example,dc=test"
[01/Sep/2023:10:25:17.347727467 -0400] - DEBUG - NSACLPlugin - print_access_control_summary - conn=59 op=5 (main): Allow search on entry(krbprincipalname=cifs/vm-10-0-185-32.hosted.upshift.rdu2.redhat.com@example.test,cn=services,cn=accounts,dc=example,dc=test).attr(objectClass) to uid=tuser1,cn=users,cn=accounts,dc=example,dc=test: cached context/parent allow
[01/Sep/2023:10:25:17.348091187 -0400] - DEBUG - NSACLPlugin - print_access_control_summary - conn=59 op=5 (on entry): Allow read on entry(krbprincipalname=cifs/vm-10-0-185-32.hosted.upshift.rdu2.redhat.com@example.test,cn=services,cn=accounts,dc=example,dc=test).attr(NULL) to uid=tuser1,cn=users,cn=accounts,dc=example,dc=test: cached context/parent allow
[01/Sep/2023:10:25:17.350074076 -0400] - DEBUG - NSACLPlugin - acl_access_allowed - #### conn=59 op=6 binddn="uid=tuser1,cn=users,cn=accounts,dc=example,dc=test"
[01/Sep/2023:10:25:17.350648602 -0400] - DEBUG - NSACLPlugin - print_access_control_summary - conn=59 op=6 (main): Deny write on entry(krbprincipalname=cifs/vm-10-0-185-32.hosted.upshift.rdu2.redhat.com@example.test,cn=services,cn=accounts,dc=example,dc=test).attr(memberPrincipal) to uid=tuser1,cn=users,cn=accounts,dc=example,dc=test: no aci matched the subject by aci(22): aciname= "Admin can manage any entry", acidn="dc=example,dc=test"

# klist
Ticket cache: KCM:0:23917
Default principal: tuser1@EXAMPLE.TEST

I'm using the same tree for testing a different PR, which I installed the server under. So this testing is on an upgraded server.

abbra commented:
I don't see that acl used at all.

Here is my test:
```
[root@master ~]# ipa-ldap-updater /usr/share/ipa/updates/73-service-rbcd.update
Update complete
The ipa-ldap-updater command was successful
[root@master ~]# ipa group-add testgroup
-----------------------
Added group "testgroup"
-----------------------
  Group name: testgroup
  GID: 78800004
[root@master ~]# ipa group-add-member testgroup --users=testuser
  Group name: testgroup
  GID: 78800004
  Member users: testuser
-------------------------
Number of members added 1
-------------------------
[root@master ~]# ipa service-add foo/master.ipa.test
--------------------------------------------
Added service "foo/master.ipa.test@IPA.TEST"
--------------------------------------------
  Principal name: foo/master.ipa.test@IPA.TEST
  Principal alias: foo/master.ipa.test@IPA.TEST
  Managed by: master.ipa.test
[root@master ~]# ipa service-allow-add-delegation foo/master.ipa.test --groups testgroup
  Principal name: foo/master.ipa.test@IPA.TEST
  Principal alias: foo/master.ipa.test@IPA.TEST
  Managed by: master.ipa.test
  Groups allowed to add resource delegation: testgroup
-------------------------
Number of members added 1
-------------------------
[root@master ~]# ipa passwd testuser
New Password: 
Enter New Password again to verify: 
----------------------------------------
Changed password for "testuser@IPA.TEST"
----------------------------------------
[root@master ~]# kdestroy -A
[root@master ~]# kinit testuser
Password for testuser@IPA.TEST: 
Password expired.  You must change it now.
Enter new password: 
Enter it again: 
[root@master ~]# ipa service-add-delegation foo/master.ipa.test HTTP/master.ipa.test
-------------------------------------------------------------------------------------
Added new resource delegation to the service principal "foo/master.ipa.test@IPA.TEST"
-------------------------------------------------------------------------------------
  Principal name: foo/master.ipa.test@IPA.TEST
  Delegation principal: HTTP/master.ipa.test@IPA.TEST
[root@master ~]# ipa service-show foo/master.ipa.test
  Principal name: foo/master.ipa.test@IPA.TEST
  Principal alias: foo/master.ipa.test@IPA.TEST
  Delegation principal: HTTP/master.ipa.test@IPA.TEST
  Keytab: False
  Managed by: master.ipa.test

```

rcritten commented:
It's because 73-service-rbcd.update wasn't packaged so I don't have the updated ACIs. I'm not sure why yet.

rcritten commented:
It's missing from Makefile.am

rcritten commented:
My problem is resolved after manually loading the update.

abbra commented:
Thank you. Added to the makefile.

rcritten commented:
That fixed it. Changes LGTM.

flo-renaud commented:
master:

* f123b01d81696c52e9a4008d46e549864e4a8069 updates: add ACIs for RBCD self-management

flo-renaud commented:
Hi @abbra 
the patch failed to apply cleanly on ipa-4-10, can you do a manual backport?
```
Applying to ipa-4-10: updates: add ACIs for RBCD self-management
Failure to apply patches: .git/rebase-apply/patch:17: new blank line at EOF.
+
warning: 1 line adds whitespace errors.
error: Failed to merge in the changes.
hint: Use 'git am --show-current-patch=diff' to see the failed patch

Failed to apply patches onto origin/ipa-4-10. Manual backport is needed.
```
